src/server/pfs/server/trigger.go

```diff
@@ -16,86 +15,73 @@ import (
-	headBranches[newHead.Commit.Branch.Name] = true
-	for _, b := range repoInfo.Branches {
-		bi := &pfs.BranchInfo{}
-		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(pfsdb.BranchKey(b), bi); err != nil {
-			return nil, err
-		}
-		if bi.Head != nil && bi.Head.ID == commit.ID {
-			headBranches[b.Name] = true
-		}
+		return err
 	}
-	triggeredBranches := map[string]bool{}
-	var result []*pfs.Branch
-	var triggerBranch func(branch *pfs.Branch) error
-	triggerBranch = func(branch *pfs.Branch) error {
-		if triggeredBranches[branch.Name] {
-			return nil
+	commitInfo := &pfs.CommitInfo{}
+	if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(pfsdb.CommitKey(commit), commitInfo); err != nil {
+		return err
+	}
+
+	// Track any branches that are triggered and their new (alias) head commit
+	triggeredBranches := map[string]*pfs.CommitInfo{}
+	triggeredBranches[commitInfo.Commit.Branch.Name] = commitInfo
+
+	var triggerBranch func(branch *pfs.Branch) (*pfs.CommitInfo, error)
+	triggerBranch = func(branch *pfs.Branch) (*pfs.CommitInfo, error) {
+		if newHead, ok := triggeredBranches[branch.Name]; ok {
+			return newHead, nil
 		}
-		triggeredBranches[branch.Name] = true
+
 		bi := &pfs.BranchInfo{}
 		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(pfsdb.BranchKey(branch), bi); err != nil {
-			return err
+			return nil, err
 		}
+
+		triggeredBranches[branch.Name] = nil
 		if bi.Trigger != nil {
-			if err := triggerBranch(commit.Branch.Repo.NewBranch(bi.Trigger.Branch)); err != nil && !col.IsErrNotFound(err) {
-				return err
+			newHead, err := triggerBranch(commit.Branch.Repo.NewBranch(bi.Trigger.Branch))
+			if err != nil && !col.IsErrNotFound(err) {
+				return nil, err
 			}
-			if headBranches[bi.Trigger.Branch] {
-				var oldHead *pfs.CommitInfo
-				if bi.Head != nil {
-					oldHead = &pfs.CommitInfo{}
-					if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(pfsdb.CommitKey(bi.Head), oldHead); err != nil {
-						return err
-					}
+
+			if newHead != nil {
+				oldHead := &pfs.CommitInfo{}
+				if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(pfsdb.CommitKey(bi.Head), oldHead); err != nil {
+					return nil, err
 				}
+
 				triggered, err := d.isTriggered(txnCtx, bi.Trigger, oldHead, newHead)
 				if err != nil {
-					return err
+					return nil, err
 				}
+
 				if triggered {
-					if err := d.branches.ReadWrite(txnCtx.SqlTx).Update(pfsdb.BranchKey(bi.Branch), bi, func() error {
-						bi.Head = newHead.Commit
-						return nil
-					}); err != nil {
-						return err
+					aliasCommit, err := d.aliasCommit(txnCtx, newHead.Commit, bi.Branch)
+					if err != nil {
+						return nil, err
+					}
+					triggeredBranches[branch.Name] = aliasCommit
+					if err := txnCtx.PropagateBranch(bi.Branch); err != nil {
+						return nil, err
 					}
-					result = append(result, proto.Clone(commit.Branch).(*pfs.Branch))
-					headBranches[bi.Branch.Name] = true
 				}
 			}
 		}
-		return nil
+		return nil, nil
 	}
 	for _, b := range repoInfo.Branches {
```

Comment:
Not related to this PR, but this seems like a scalability problem for branches (noteworthy if we want to go the branch route for run pipeline).